From dcaeda81fed54f75ad49eb781c7ac852baf4712d Mon Sep 17 00:00:00 2001
From: Alexandre Besnard <alexandre.besnard@softathome.com>
Date: Wed, 26 Jun 2024 15:54:06 +0200
Subject: [PATCH 1/2] Declare engine pkey variable in its own code block

---
 lib/net_mosq.c | 4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/lib/net_mosq.c b/lib/net_mosq.c
index b8d14a02..0192dc5b 100644
--- a/lib/net_mosq.c
+++ b/lib/net_mosq.c
@@ -641,9 +641,6 @@ static int net__init_ssl_ctx(struct mosquitto *mosq)
 	ENGINE *engine = NULL;
 	uint8_t tls_alpn_wire[256];
 	uint8_t tls_alpn_len;
-#if !defined(OPENSSL_NO_ENGINE)
-	EVP_PKEY *pkey;
-#endif
 
 #ifndef WITH_BROKER
 	if(mosq->user_ssl_ctx){
@@ -784,6 +781,7 @@ static int net__init_ssl_ctx(struct mosquitto *mosq)
 			if(mosq->tls_keyfile){
 				if(mosq->tls_keyform == mosq_k_engine){
 #if !defined(OPENSSL_NO_ENGINE)
+					EVP_PKEY *pkey;
 					UI_METHOD *ui_method = net__get_ui_method();
 					if(mosq->tls_engine_kpass_sha1){
 						if(!ENGINE_ctrl_cmd(engine, ENGINE_SECRET_MODE, ENGINE_SECRET_MODE_SHA, NULL, NULL, 0)){
-- 
2.45.2

